File: compiler/nodes/expressions/placeholder.go
package expressions

import (
	"wipple/codegen"
	"wipple/database"
	"wipple/syntax"
	"wipple/visit"
)

type IsPlaceholder struct{}

func (fact IsPlaceholder) String() string {
	return "is a placeholder"
}

type PlaceholderExpressionNode struct {
	Facts *database.Facts
}

func (node *PlaceholderExpressionNode) GetFacts() *database.Facts {
	return node.Facts
}

func ParsePlaceholderExpression(parser *syntax.Parser) (*PlaceholderExpressionNode, *syntax.Error) {
	span := parser.Spanned()

	_, err := parser.Token("UnderscoreKeyword")
	if err != nil {
		return nil, err
	}

	parser.Commit("in this placeholder expression")

	return &PlaceholderExpressionNode{
		Facts: database.NewFacts(span()),
	}, nil
}

func (node *PlaceholderExpressionNode) Visit(visitor *visit.Visitor) {
	visitExpression(visitor, node)

	database.SetFact(node, IsPlaceholder{})
}

func (node *PlaceholderExpressionNode) Codegen(c *codegen.Codegen) error {
	return c.Error(node)
}

File: compiler/src/nodes/expressions/placeholder.rs
use crate::{
    codegen::{Codegen, CodegenCtx, ir},
    database::{Fact, Node, NodeRef, Render},
    nodes::visit_expression,
    syntax::{ParseError, Parser, TokenKind},
    visit::{Visit, Visitor},
};

#[derive(Debug, Clone)]
pub struct IsPlaceholder;

impl Fact for IsPlaceholder {}

impl Render for IsPlaceholder {}

#[derive(Debug)]
pub struct PlaceholderExpressionNode;

impl Node for PlaceholderExpressionNode {}

pub fn parse_placeholder_expression(
    parser: &mut Parser<'_>,
) -> Result<PlaceholderExpressionNode, ParseError> {
    parser.token(TokenKind::UnderscoreKeyword)?;
    parser.commit("in this placeholder expression");

    Ok(PlaceholderExpressionNode)
}

impl Visit for PlaceholderExpressionNode {
    fn visit(&self, node: &NodeRef, visitor: &mut Visitor<'_>) {
        visit_expression(node, visitor);
        visitor.insert(node, IsPlaceholder);
    }
}

impl Codegen for PlaceholderExpressionNode {
    fn codegen(&self, _node: &NodeRef, _ctx: &mut CodegenCtx<'_>) -> Option<ir::SpannedExpression> {
        None
    }
}
